### ISSUE TITLE
Parallel index creation on distributed tables?

### ISSUE THREAD

**Issue by colton-citus** [citusdata/citus#2408]
If you connect to the coordinator using multiple simultaneous connections and issue a series of CREATE INDEX statements on the same table, the indexes will run in series, where index B only starts being created on all workers once index A is finished on all workers. 

The PostgreSQL 10 documentation makes it clear that it is possible to create multiple indexes on the same table in parallel: 

https://www.postgresql.org/docs/current/static/sql-createindex.html 

This issue is being opened to clarify expected behavior and/or request parallel index creation if this is not currently expected behavior.

**Comment by marcocitus** [comment]
`CREATE INDEX` runs in parallel across all shard placements. `CREATE INDEX `CONCURRENTLY` does not, in part because Postgres does not allow `CREATE INDEX CONCURRENTLY` to run in parallel.

We could run `CREATE INDEX CONCURRENTLY` in parallel across nodes, but currently don't due to executor limitations.

Should we rename this issue to "Parallelise CREATE INDEX CONCURRENTLY across nodes"?

**Comment by colton-citus** [comment]
I haven't been able to get CREATE INDEX to run in parallel across shard placements in my testing so far- all the workers have finished creating the first index before starting on the second. I'll do some further testing to make sure that this isn't a problem with my environment and nail down reproduction steps and update shortly. 

I am not using CONCURRENTLY in my testing.

**Comment by marcocitus** [comment]
How are you checking this? I cannot reproduce a case where the CREATE INDEX is not executed parallel.

**Comment by seamusabshere** [comment]
@marcocitus I initially reported this bug by running `CREATE INDEX` on 3 separate connections. On the coordinator, it shows up as them all running.

If you look at the worker nodes, though, only one set of `CREATE INDEX` run at a time.

**Comment by seamusabshere** [comment]
@marcocitus here's an example from literally right now, citus cloud with citus 7.5 and postgres 10.4

each worker node is shown with all of its active queries

<img width="855" alt="screen shot 2018-10-17 at 4 44 51 pm" src="https://user-images.githubusercontent.com/15787/47115438-91203080-d22c-11e8-9f1d-0cb20a40af54.png">

**Comment by marcocitus** [comment]
We had some confusion around the meaning of parallel.

Internally, we tend to use:
Parallel CREATE INDEX - CREATE INDEX is executed in parallel across shards 
Concurrent CREATE INDEX - Multiple CREATE INDEX statements issued by the client concurrently

I was thinking you were saying the first one doesn't work, which it does, but as you're rightly pointing out Citus only executes one CREATE INDEX at a time (at least, on the same table).

The reason is we have some locks on multi-shard commands to prevent multi-shard commands from running into distributed deadlocks  a lot (e.g. conflicting updates that execute on shards in different order). I think that we don't really need them for CREATE INDEX or DDL commands in general. It might be worth avoiding those locks, though I'm not sure whether it'll actually increase machine utilisation, since the parallelisation already makes a single CREATE INDEX use many cores and issuing additional commands might just cause contention.

**Comment by seamusabshere** [comment]
> issuing additional commands might just cause contention

right, but i think it should be up to the worker node's postgres to decide that - citus shouldn't add a global lock on CREATE INDEX unless it's necessary IMO

**Comment by seamusabshere** [comment]
@marcocitus is it possible that this same issue affects indexes on reference tables, effectively only allowing one node to make indexes at a time?

**Comment by ehmo** [comment]
Guys any progress on running the `CONCURRENTLY` index creation across all the nodes? It's incredibly inefficient when it runs on a single node at a time, especially for a big table. The longest operation took literally 3 days.
